src/Common/PhpTypesTrait.php

```diff
@@ -9,18 +9,41 @@
 
 trait PhpTypesTrait
 {
+    /**
+     * @param string $type
+     *
+     * @return string
+     */
+    protected function stripNullReturnTypeHint(string $type): string
+    {
+        return str_replace('|null', '', $type);
+    }
+
     /**
      * @param string $type
      *
      * @return bool
      */
     protected function isTypeInPhp7NotAllowed(string $type): bool
     {
-        $type = str_replace('|null', '', $type);
+        $type = $this->stripNullReturnTypeHint($type);
 
         return strpos($type, '|') !== false
             || $type === 'mixed'
             || $type === 'false'
             || $type === 'static';
     }
+
+    /**
+     * @param string $type
+     *
+     * @return bool
+     */
+    protected function isReturnTypeIsThis(string $type): bool
+    {
+        $type = $type = $this->stripNullReturnTypeHint($type);
+
+        return strpos($type, '|') !== false
+            || $type === '$this';
```

Comment:
This is not respecting also other types that should not be added as type, as discussed
As such, the method name is also not good here.